server/etcdserver/api/etcdhttp/health.go

```diff
@@ -355,8 +356,16 @@ func activeAlarmCheck(srv ServerHealth, at pb.AlarmType) func(context.Context) e
 	}
 }
 
-func serializableReadCheck(srv ServerHealth) func(ctx context.Context) error {
+func serializableReadCheck(srv ServerHealth, skipCheckDuringDefrag bool) func(ctx context.Context) error {
 	return func(ctx context.Context) error {
+		// skips the check if defrag is active.
+		if skipCheckDuringDefrag && srv.IsDefragActive() {
```

Comment:
What if there is a deadlock on defrag? I think I missed this in the original design, but I don't think we should just skip checking the serializable read if defrag is active. Single liveness probe failure during defrag is not a bad thing. We expect that 3 failures are required to kill etcd.

The question is not "should defrag cause livez to fail?", but "should we kill etcd if defrag is taking 15 seconds?".
I think yes.

Reply:
I guess if the user expects the defrag latency to be long, they should probably configure the livez retry period to be long to match that.

Reply:
> If we allow livez to fail during defrag, this etcd process will be in a crash loop state, right?

No crash loop as defrag operation will not persist between restarts. 

> In cloud environment, it is not uncommon that defrag could take more than 15 seconds with large number of key values and low throughput disk volume.

Asked @siyuanfoundation to take a look into defrag duration metrics. I agree that defrag can take couple of seconds, however I think 15s of API unavailability seems too long. I think we should assume some minimal IOPs from disk size. 

You are right it possible that on cloud users could provision minimal disk sizes (20GB) to just fit etcd db (6GB) and this will result with minimal IOPs, however I would argue that those are degenerate cases and those clusters would handle etcd benchmark. 

It's hard for me to judge what is expected defrag duration I don't think there is any official SLO or set of it. It  was not a problem for GKE so I haven't investigated further. I think the best way to get an agreement would be to share some internal data about defrag duration. 

@chaochn47 I know you were interested in it due to grpc probes. Could you share some data?

Reply:
I double checked the defrag duration stats. I have confirmed that defrag time can be really long on large clusters, even with over provisioned disks it can grow easily over 15s. As long as defrag time is linear to size of DB, we won't be able to define a sensible maximum time that we can enforce.

With that I would agree that the proposal makes sense.